### ISSUE TITLE
gradient difference between single GPU and multi-GPU DataParallel

### ISSUE THREAD

**Issue by haipeng00** [pytorch/pytorch#14007]
Running the following scripts with pytorch0.4.1, cuda8 and cudnn5, gives different gradients between single GPU and multi GPU. (scripts quoted from https://github.com/pytorch/pytorch/issues/5778#issuecomment-373086192)

```python
import torch
input = torch.randn(4, 128, 128).cuda()
linear = torch.nn.Linear(128, 128).cuda()
dp = torch.nn.DataParallel(linear, [0, 1])

linear(input).sum().backward()
print(linear.weight.grad)

linear.zero_grad()
dp(input).sum().backward()
print(linear.weight.grad)
```

Single-GPU gradients:
tensor([[ 38.8789596558, -21.8231239319,  24.4459419250,  ...,
          16.9406051636,  16.2471446991, -39.3904113770],
        [ 38.8789596558, -21.8231239319,  24.4459419250,  ...,)

Multi-gpu DataParallel gradients:
tensor([[ 38.8789596558, -21.8231220245,  24.4459438324,  ...,
          16.9406070709,  16.2471446991, -39.3904151917],
        [ 38.8789596558, -21.8231220245,  24.4459438324,  ...,)

There are tiny differences between the gradients.  Any idea where does the difference come from? Thanks a lot!

**Comment by soumith** [comment]
the order of doing operations accumulates floating point errors in different ways. This is the reason for this difference, and it's expected.
